File: src/main/java/miragefairy2019/mod/modules/fairycrystal/VariantFairyCrystal.java
package miragefairy2019.mod.modules.fairycrystal;

import miragefairy2019.libkt.WeightedItem;
import miragefairy2019.libkt.WeightedItemKt;
import miragefairy2019.mod3.skill.api.ApiSkill;
import mirrg.boron.util.suppliterator.ISuppliterator;
import net.minecraft.entity.item.EntityItem;
import net.minecraft.entity.player.EntityPlayer;
import net.minecraft.item.ItemStack;
import net.minecraft.stats.StatList;
import net.minecraft.util.EnumActionResult;
import net.minecraft.util.EnumFacing;
import net.minecraft.util.EnumHand;
import net.minecraft.util.math.BlockPos;
import net.minecraft.util.text.ITextComponent;
import net.minecraft.util.text.TextComponentString;
import net.minecraft.world.World;
import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.Optional;

public class VariantFairyCrystal extends VariantFairyCrystalBase {

    public VariantFairyCrystal(@NotNull String registryName, @NotNull String unlocalizedName, @NotNull String oreName) {
        super(registryName, unlocalizedName, oreName);
    }

    public EnumActionResult onItemUse(EntityPlayer player, World world, BlockPos pos, EnumHand hand, EnumFacing facing, float hitX, float hitY, float hitZ) {
        ItemStack itemStackCrystal = player.getHeldItem(hand);
        if (itemStackCrystal.isEmpty()) return EnumActionResult.PASS;

        if (!player.isSneaking()) {
            if (world.isRemote) return EnumActionResult.SUCCESS;

            // ガチャを引く
            Optional<ItemStack> oItemStack = getDropper().drop(player, world, pos, hand, facing, hitX, hitY, hitZ, getDropRank(), getRareBoost(ApiSkill.skillManager.getServerSkillContainer(player)));

            // ガチャが成功した場合、
            if (oItemStack.isPresent()) {
                if (!oItemStack.get().isEmpty()) {

                    // ガチャアイテムを消費
                    itemStackCrystal.shrink(1);
                    player.addStat(StatList.getObjectUseStats(itemStackCrystal.getItem()));

                    // 妖精をドロップ
                    BlockPos pos2 = pos.offset(facing);
                    EntityItem entityitem = new EntityItem(world, pos2.getX() + 0.5, pos2.getY() + 0.5, pos2.getZ() + 0.5, oItemStack.get().copy());
                    entityitem.setNoPickupDelay();
                    world.spawnEntity(entityitem);

                }
            }

            return EnumActionResult.SUCCESS;
        } else {
            if (world.isRemote) return EnumActionResult.SUCCESS;

            // ガチャリスト取得
            List<WeightedItem<ItemStack>> dropTable = getDropper().getDropTable(player, world, pos, hand, facing, hitX, hitY, hitZ, getDropRank(), getRareBoost(ApiSkill.skillManager.getServerSkillContainer(player)));

            // 表示
            ITextComponent string = new TextComponentString("");
            string.appendText("===== " + itemStackCrystal.getDisplayName() + " (" + (world.isRemote ? "Client" : "Server") + ") =====");
            string.appendText("\n");
            double totalWeight = WeightedItemKt.getTotalWeight(dropTable);
            for (WeightedItem<ItemStack> weightedItem : ISuppliterator.ofIterable(dropTable)
                .sortedObj(i -> i.getItem().getDisplayName())
                .sortedDouble(i -> i.getWeight())) {
                string.appendText(String.format("%f%%", 100 * weightedItem.getWeight() / totalWeight) + ": ");
                string.appendText(weightedItem.getItem().getDisplayName());
                string.appendText("\n");
            }
            string.appendText("====================");
            player.sendStatusMessage(string, false);

            return EnumActionResult.SUCCESS;
        }

    }

}

File: src/main/java/miragefairy2019/mod/modules/fairycrystal/VariantFairyCrystal.kt
package miragefairy2019.mod.modules.fairycrystal

import miragefairy2019.libkt.drop
import miragefairy2019.libkt.orNull
import miragefairy2019.libkt.textComponent
import miragefairy2019.libkt.totalWeight
import miragefairy2019.mod3.skill.api.ApiSkill
import mirrg.kotlin.formatAs
import net.minecraft.entity.player.EntityPlayer
import net.minecraft.stats.StatList
import net.minecraft.util.EnumActionResult
import net.minecraft.util.EnumFacing
import net.minecraft.util.EnumHand
import net.minecraft.util.math.BlockPos
import net.minecraft.util.math.Vec3d
import net.minecraft.util.text.ITextComponent
import net.minecraft.world.World

open class VariantFairyCrystal(registryName: String, unlocalizedName: String, oreName: String) : VariantFairyCrystalBase(registryName, unlocalizedName, oreName) {
    fun onItemUse(player: EntityPlayer, world: World, pos: BlockPos, hand: EnumHand, facing: EnumFacing, hitX: Float, hitY: Float, hitZ: Float): EnumActionResult {
        val crystalItemStack = player.getHeldItem(hand).orNull ?: return EnumActionResult.PASS // アイテムが消えた場合は中止
        if (world.isRemote) return EnumActionResult.SUCCESS // サーバースレッドのみ

        val rareBoost = getRareBoost(ApiSkill.skillManager.getServerSkillContainer(player))

        if (!player.isSneaking) {

            // ガチャを引く
            val resultItemStack = dropper.drop(player, world, pos, hand, facing, hitX, hitY, hitZ, dropRank, rareBoost)?.orNull ?: return EnumActionResult.SUCCESS

            // ガチャ成立

            // ガチャアイテムを消費
            crystalItemStack.shrink(1)
            StatList.getObjectUseStats(crystalItemStack.item)?.let { player.addStat(it) }

            // 妖精をドロップ
            val pos2 = pos.offset(facing)
            resultItemStack.drop(world, Vec3d(pos2).addVector(0.5, 0.5, 0.5), noPickupDelay = true)

            return EnumActionResult.SUCCESS
        } else {

            // ガチャリスト取得
            val dropTable = dropper.getDropTable(player, world, pos, hand, facing, hitX, hitY, hitZ, dropRank, rareBoost)


            // 表示

            fun send(textComponent: ITextComponent) = player.sendStatusMessage(textComponent, false)

            send(textComponent { !"===== " + !crystalItemStack.displayName + !" (${if (world.isRemote) "Client" else "Server"}) =====" })

            val totalWeight = dropTable.totalWeight
            dropTable.sortedBy { it.item.displayName }.sortedBy { it.weight }.forEach { weightedItem ->
                send(textComponent { !"${weightedItem.weight / totalWeight * 100.0 formatAs "%f%%"}: " + !weightedItem.item.displayName })
            }

            send(textComponent { !"====================" })


            return EnumActionResult.SUCCESS
        }
    }
}
